### PR TITLE
BUG: Allow basic sequences to be multipled with bool_

### LINKED ISSUE #9234: 'string'*numpy.bool_(True) raises TypeError; was ok in numpy 0.9
Dear all

I encountered a potential regression when upgrading from numpy 1.9 to 1.12.

```python
>>> 'string'*numpy.bool_(True) 
TypeError: an integer is required
```

Numpy version: 1.12.1
Python 2.7.10 (default, May 23 2015, 09:44:00) [MSC v.1500 64 bit (AMD64)]
Windows 7, 64-bit

This worked ok in Numpy 1.9 with the same Python and OS version.

Actually, this is is simplified from the actual use case:
```python
>>>a_np = numpy.float64(3.1415)
>>>'string'*(a_np>0)
```

This works ok when a_np is numpy.float(3.1415), because the result type of the comparison is bool, not numpy.bool_.

Thank you and best regards
Martin

### PR DESCRIPTION
**Pull request by kitchoi**
Closes #9234

### CODE REVIEW

**Comment by charris** [comment]
Off hand, I don't think this is a good idea.

**Comment by kitchoi** [comment]
I was trying to implement what was suggested in the latest discussion in #9234, which has been identified as a bug.  Did I misunderstand the issue?

**Comment by kitchoi** [comment]
My understanding of the issue is that the error in #9234 was accidental.
If `"string"*numpy.True_` should fail, then it should fail because `bool_.__index__` should fail via cpython's `sequence_repeat`.  `bool_.__index__` is now going through a deprecation process (https://github.com/numpy/numpy/issues/7395).  Eventually `"string"*numpy.True_` will fail with an error coming from `bool_.__index__`.

**Comment by eric-wieser** [comment]
I suppose I'd argue that if it's been 4(?) releases until we realized we skipped a deprecation cycle, there's little point in introducing it now.

**Comment by kitchoi** [comment]
An alternative would be to apply the change once `bool_.__index__` is removed?
I'd argue that the change is still justified by the fact that it follows Python's sequence multiplication:
```
class IndexLike:

    def __index__(self):
        return 3


class NotAnIndex:
    pass

index = IndexLike()
s = "string"
print(s*index)    # stringstringstring

x = [1, 2, 3]
print(x*index)   # [1, 2, 3, 1, 2, 3, 1, 2, 3]

not_index = NotAnIndex()
# TypeError
# Python 3.5: TypeError: can't multiply sequence by non-int of type 'NotAnIndex'
# Python 2.7: TypeError: object cannot be interpreted as an index
print(x*not_index)
```

**Comment by eric-wieser** [comment]
Ok, I'm back on board for this, as it makes the error slightly closer to how python behaves.

Really though, we should match the sequence implementation of `PyNumber_Multiply`, which looks like:

```
static PyObject *
sequence_repeat(ssizeargfunc repeatfunc, PyObject *seq, PyObject *n)
{
    Py_ssize_t count;
    if (PyIndex_Check(n)) {
        count = PyNumber_AsSsize_t(n, PyExc_OverflowError);
        if (count == -1 && PyErr_Occurred())
            return NULL;
    }
    else {
        return type_error("can't multiply sequence by "
                          "non-int of type '%.200s'", n);
    }
    return (*repeatfunc)(seq, count);
}
```

It's not entirely clear to me why we can't just return `NotImplemented` in `nb_multiply`, and let python handle things.

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 1]
```diff
@@ -271,7 +271,7 @@ gentype_multiply(PyObject *m1, PyObject *m2)
             ((Py_TYPE(m2)->tp_as_number == NULL) ||
              (Py_TYPE(m2)->tp_as_number->nb_multiply == NULL))) {
         /* Try to convert m1 to an int and try sequence repeat */
-        repeat = PyArray_PyIntAsIntp(m1);
+        repeat = PyNumber_AsSsize_t(m1, PyExc_OverflowError);
         if (error_converting(repeat)) {
             return NULL;
         }
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/scalartypes.c.src]: Simply using the following for this whole chunk seems to work correctly:
```
        Py_INCREF(Py_NotImplemented);
        return Py_NotImplemented;
```

Comment (eric-wieser) [reply numpy/core/src/multiarray/scalartypes.c.src]: The only test that fails is the one that fails in this patch, because the warning needs silencing.

Comment (kitchoi) [reply numpy/core/src/multiarray/scalartypes.c.src]: Yes I agree that returning NotImplemented is better.

**Comment by kitchoi** [comment]
Follow up on @eric-wieser comment on the fact that the error has been there for multiple releases, out of curiosity, I found that the error indeed occurs since 1.12.0.  Furthermore there is in fact a deprecation warning message in 1.11.0 from https://github.com/numpy/numpy/blob/v1.11.0/numpy/core/src/multiarray/conversion_utils.c#L794-L822

So the usage has in fact had its deprecation cycle (if you trust my somewhat manual investigation).  Now I am more inclined to applying this (or similar) change after `bool_.__index__` is gone. :)  If so it might make sense to close this since nearly half of the changes would not be applicable.  What do people think?

**Pull request closed without merging**